pinot-core/src/main/java/org/apache/pinot/core/data/manager/realtime/IngestionDelayTracker.java

```diff
@@ -0,0 +1,350 @@
+      RealtimeTableDataManager tableDataManager) {
+    this(serverMetrics, metricNamePrefix + tableNameWithType, tableDataManager,
+        TIMER_THREAD_TICK_INTERVAL_MS, metricNamePrefix, true, true);
+  }
+
+
+  /**
+   * Use to set or reset the aging of reported values.
+   * @param enableAging true if we want maximum to be aged as per sample time or false if we do not want to age
+   *                   samples
+   */
+  @VisibleForTesting
+  void setEnableAging(boolean enableAging) {
+    _enableAging = enableAging;
+  }
+
+  /*
+   * Called by LLRealTimeSegmentDataManagers to post delay updates to this tracker class.
+   * If the new sample represents a new Maximum we update the current maximum.
+   * If the new sample was for the partition that was maximum, but delay is not maximum anymore, we must select
+   * a new maximum.
+   *
+   * @param delayInMilliseconds ingestion delay being recorded.
+   * @param sampleTime sample time.
+   * @param partitionGroupId partition ID for which this delay is being recorded.
+   */
+  public void updateIngestionDelay(long delayInMilliseconds, long sampleTime, int partitionGroupId) {
+    // Store new measure and wipe old one for this partition
+    DelayMeasure previousMeasure = _partitionToDelaySampleMap.put(partitionGroupId,
+        new DelayMeasure(sampleTime, delayInMilliseconds));
+    if ((previousMeasure == null) && _enablePerPartitionMetric) {
+      // First time we start tracking a partition we should start tracking it via metric
+      _serverMetrics.addCallbackTableGaugeIfNeeded(getPerPartitionMetricName(partitionGroupId),
+          ServerGauge.TABLE_PER_PARTITION_INGESTION_DELAY_MS,
+          () -> getPartitionIngestionDelay(partitionGroupId));
+    }
+    // If we are consuming we do not need to track this partition for removal.
+    _partitionsMarkedForVerification.remove(partitionGroupId);
+  }
+
+  /*
+   * Handle partition removal event. This must be invoked when we stop serving a given partition for
+   * this table in the current server.
+   * This function will be invoked when we receive CONSUMING -> DROPPED / OFFLINE state transitions.
+   *
+   * @param partitionGroupId partition id that we should stop tracking.
+   */
+  public void stopTrackingPartitionIngestionDelay(int partitionGroupId) {
+    removePartitionId(partitionGroupId);
+  }
+
+  /*
+   * This method is used for timing out inactive partitions, so we don't display their metrics on current server.
+   * When the inactive time exceeds some threshold, we read from ideal state to confirm we still host the partition,
+   * if not we remove the partition from being tracked locally.
+   * This call is to be invoked by a timer thread that will periodically wake up and invoke this function.
+   */
+  public void timeoutInactivePartitions() {
+    List<Integer> partitionsHostedByThisServer = null;
+    try {
+      partitionsHostedByThisServer = getPartitionsHostedByThisServerPerIdealState();
```

Comment:
Important: Do not make this call if there are no partitions marked for verification (mostly the case)

Reply:
Good optimization, let me address this right away

Reply:
Done in the las commit